plugins/filter_watermark/watermark.c

```diff
@@ -0,0 +1,392 @@
+                              const char *tag, int tag_len,
+                              void **out_buf, size_t *out_size,
+                              struct flb_filter_instance *f_ins,
+                              void *context,
+                              struct flb_config *config)
+{
+    msgpack_object root;
+    msgpack_object map;
+    msgpack_object key;
+    msgpack_object value;
+    int i, map_size, ret;
+    struct tm tm0;
+
+    msgpack_object root_tmp;
+    msgpack_unpacked result_tmp;
+    size_t off_tmp = 0;
+    size_t inner_bytes = 0;
+    
+    msgpack_unpacked result;
+    msgpack_unpacked *destroy_result;
+    size_t off = 0;
+
+    char tmp_buf[45];
+    int tmp_buf_len = 45;
+
+    msgpack_sbuffer mp_sbuf;
+    msgpack_packer mp_pck;
+    msgpack_sbuffer *tmp_sbuf;
+    
+    struct flb_filter_watermark_ctx *ctx = context; 
+    destroy_result = &result;
+
+    msgpack_sbuffer_init(&mp_sbuf);
+    msgpack_packer_init(&mp_pck, &mp_sbuf, msgpack_sbuffer_write);
+
+    tmp_sbuf = msgpack_sbuffer_new();
+    msgpack_sbuffer_init(tmp_sbuf);
+    
+
+    msgpack_unpacked_init(&result);
+    while (msgpack_unpack_next(&result, data, bytes, &off) == MSGPACK_UNPACK_SUCCESS) {
+        root = result.data;
+        map = root.via.array.ptr[1];
+        map_size = map.via.map.size;
+
+        for (i = 0; i < map_size; i++) {
+            key = map.via.map.ptr[i].key;
+
+            if (flb_sds_cmp(ctx->time_field, (char *) key.via.str.ptr, key.via.str.size) != 0) {
+                 /* ignore the key if the key is not the same as time_filed. */
+                 continue;
+            }
+
+            value = map.via.map.ptr[i].val;
+            if (value.type == MSGPACK_OBJECT_STR) {
+                flb_plg_debug(ctx->ins, "size is %d", value.via.str.size);
+                flb_plg_debug(ctx->ins, "content is %s", value.via.str.ptr);
+ 		
+		if(value.via.str.size > tmp_buf_len) {
+                   flb_plg_info(ctx->ins, "Timestamp string length is too long to be correct, drop the record\n");
+                   continue;
+		}
+
+                flb_sds_t buf;
+                buf = flb_sds_create_len(value.via.str.ptr, value.via.str.size);
+		memset(tmp_buf, 0, tmp_buf_len);
+                strcpy(tmp_buf, buf);
+                flb_sds_destroy(buf);
+
+            }
+
+            /* we reply on strptime to check the format. */
+            memset(&tm0, 0, sizeof(struct tm));
+            /* strptime("2021-11-12 18:31:01", "%Y-%m-%d %H:%M:%S", &tm0); */
+            if( strptime(tmp_buf, "%Y-%m-%d %H:%M:%S", &tm0) == 0) {
+                flb_plg_info(ctx->ins, "Timestamp format is not acceptable by strptime, drop the record\n");
+		continue;
+            }
+
+           /* when it reaches here, we need to malloc memory to store all the content of record and insert record into heap. */
+	   struct flb_filter_watermark_record *tmp_ptr = flb_malloc(sizeof(struct flb_filter_watermark_record));
```

Comment:
Variables must be declared at the beginning of a function and not in the middle of the code (see guidelines).